### PR TITLE
#30468 parse supported timezone

### PR DESCRIPTION
**Pull request by suneox**
### Fixed Issues

$ https://github.com/Expensify/App/issues/30468
PROPOSAL: https://github.com/Expensify/App/issues/30468#issuecomment-1782518130

### Tests
Preconditions: simulate location or change timezone to Ukraine
  - Web config: Config `Europe/Kiev` [location](https://developer.chrome.com/docs/devtools/settings/locations) then turn on [sensor](https://developer.chrome.com/docs/devtools/sensors#geolocation)
  - IOS config: https://www.makeuseof.com/how-to-change-time-zone-iphone/ 
  - Android config: https://support.google.com/android/answer/2841106?hl=en

1. Open Settings
2. Click on Profile
3. Click on the Timezone
4. Turn on "Automatically determine your location." and observe the timezone
5. Turn off "Automatically determine your location."
6. Click on the timezone selection
7. Verify the timezone at step 4 exists in the list

- [x] Verify that no errors appear in the JS console

### QA Steps
Preconditions: simulate location or change timezone to Ukraine
  - Web config: Config `Europe/Kiev` [location](https://developer.chrome.com/docs/devtools/settings/locations) then turn on [sensor](https://developer.chrome.com/docs/devtools/sensors#geolocation)
  - IOS config: https://www.makeuseof.com/how-to-change-time-zone-iphone/ 
  - Android config: https://support.google.com/android/answer/2841106?hl=en

1. Open Settings
2. Click on Profile
3. Click on the Timezone
4. Turn on "Automatically determine your location." and observe the timezone
5. Turn off "Automatically determine your location."
6. Click on the timezone selection
7. Verify the timezone at step 4 exists in the list

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/11959869/1a5b88d6-6b6f-414e-ae6e-45bca2427e4b

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/11959869/09182817-e4db-4813-abca-aa5c23cec7f9

</details>

<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/11959869/2c7c9dd2-3861-4777-b7bf-60a5352c6230

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/11959869/d551fd35-5062-49f9-8631-8ec09baa13fc

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/11959869/44e5b763-1e67-480d-bc8c-6d10bff43227

</details>

### CODE REVIEW

**Review by venture1981 — COMMENTED** [review]
these change work on your side?

**Comment by venture1981** [comment]
> > these change work on your side?
> 
> Yes It works I also capture in Screenshots/Videos section

Okie. great. 👍

File: src/components/withCurrentUserPersonalDetails.tsx [hunk 1]
```diff
@@ -37,7 +38,10 @@ export default function <TProps extends WithCurrentUserPersonalDetailsProps, TRe
         const accountID = props.session?.accountID ?? 0;
         const accountPersonalDetails = personalDetails?.[accountID];
         const currentUserPersonalDetails: CurrentUserPersonalDetails = useMemo(
-            () => (accountPersonalDetails ? {...accountPersonalDetails, accountID} : {}),
+            () =>
+                accountPersonalDetails
+                    ? {...accountPersonalDetails, accountID, timezone: accountPersonalDetails.timezone ? DateUtils.parseTimezone(accountPersonalDetails.timezone) : undefined}
+                    : {},
```
Comment (sobitneupane) [inline src/components/withCurrentUserPersonalDetails.tsx:44]: @suneox Can you please help me understand the reason for this change. I think we should change the timezone to supported format before saving it to Onyx. Is there any possible scenario where `personalDetails` retrieved from storage has unsupported timezone?

File: src/libs/DateUtils.ts [hunk 2]
```diff
@@ -668,6 +669,22 @@ function formatWithUTCTimeZone(datetime: string, dateFormat: string = CONST.DATE
     return '';
 }
 
+/**
+ *
+ * @param timezone
+ * function parse timezone to supported timezone
+ * @returns Timezone
+ */
+function parseTimezone(timezoneInput: Timezone): Timezone {
```
Comment (sobitneupane) [inline src/libs/DateUtils.ts:678]: I think we should name it `formatToSupportedTimezone` or `convertToSupportedTimezone`?

**Comment by sobitneupane** [comment]
@suneox Let's add the test steps for the case of App Crash (that you mention [here](https://github.com/Expensify/App/issues/30468#issuecomment-1850382655)) as well

File: src/libs/actions/PersonalDetails.ts [hunk 3]
```diff
@@ -296,7 +297,7 @@ function updateAutomaticTimezone(timezone: Timezone) {
     };
 
     const parameters: UpdateAutomaticTimezoneParams = {
-        timezone: JSON.stringify(timezone),
+        timezone: JSON.stringify(DateUtils.parseTimezone(timezone)),
```
Comment (sobitneupane) [inline src/libs/actions/PersonalDetails.ts:300]: Looks like you have not updated the timezone saved in Onyx.

https://github.com/Expensify/App/pull/33186/files#diff-1372b8700afa9b26d89dfe0575dcfaaf399d62b9b0555f30863e8c19c5609df9R310

File: src/components/withCurrentUserPersonalDetails.tsx (hunk 1 above)
Comment (sobitneupane) [reply src/components/withCurrentUserPersonalDetails.tsx:44]: Thanks @suneox for the prompt reply. I didn't get you.

> user has save the timezone is not supported to BE and user do not anything action to re-update timezone.

Can you please add more details. User saved timezone will be from the list of `TIMEZONE.ts`, right?  Will there be any case where the saved timezone is not supported by BE? Please let me know if I am missing something.

Comment (sobitneupane) [reply src/components/withCurrentUserPersonalDetails.tsx:44]: I don't think we need to consider the situation as we are still in development phase and the issue was not reported before. So, I think we are safe to remove the change. Please let me know if you disagree.

**Review by sobitneupane — COMMENTED** [review]
I am still seeing the issue @suneox 

https://github.com/Expensify/App/assets/25876548/51085ee3-a7de-4d83-b7e6-c28cad81879d

Some issues:
1. Timezone is switching between unformatted and formatted timezone when user toggles 'automatically determine location'
2. Timezone resets back to unformatted timezone.

**Review by sobitneupane — APPROVED** [review]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/25876548/ca9d6425-979a-40a5-8570-8f6e4ef7ee82

</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



https://github.com/Expensify/App/assets/25876548/7b8253f3-2709-4f7c-8ca1-9fb0b8d0e11f



</details>

<details>
<summary>iOS: mWeb Safari</summary>



https://github.com/Expensify/App/assets/25876548/f842bb47-8643-43e9-9ab4-56d5583c4637

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/25876548/360a32c1-06b3-40ce-9c59-1b1c4a70b796

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/25876548/71443d28-57a8-489d-8b0c-d6eef3ff22ff

</details>

File: src/TIMEZONES.ts [hunk 4]
```diff
@@ -419,3 +420,133 @@ export default [
     'Pacific/Wake',
     'Pacific/Wallis',
 ] as const;
+
+const timezoneBackwardMap: Record<string, (typeof TIMEZONES)[number]> = {
```
Comment (amyevans) [inline src/TIMEZONES.ts:424]: Could you add a comment explaining briefly where these values come from (link to wikipedia page)/what the purpose of the map is? That way people curious in the future don't have to follow the GH blame trail.

**Review by amyevans — CHANGES_REQUESTED** [review]
Thanks for your patience while I was OOO. Just one change requested!

File: src/TIMEZONES.ts [hunk 5]
```diff
@@ -419,3 +420,137 @@ export default [
     'Pacific/Wake',
     'Pacific/Wallis',
 ] as const;
+
+/**
+ * On some devices the timezone backward doesn't support so when a user using multiple devices has automatic timezone setting will not display in the list timezone selection
```
Comment (amyevans) [inline src/TIMEZONES.ts:425]:
```suggestion
The timezones supported in browser and on native devices differ, so we must map each timezone to its supported equivalent.
```

File: src/TIMEZONES.ts [hunk 6]
```diff
@@ -419,3 +420,137 @@ export default [
...
+ * List timezone backward parse from https://en.wikipedia.org/wiki/List_of_tz_database_time_zones
```
Comment (amyevans) [inline src/TIMEZONES.ts:426]:
```suggestion
 * Data sourced from https://en.wikipedia.org/wiki/List_of_tz_database_time_zones
```

**Review by amyevans — COMMENTED** [review]
A couple suggestions regarding the comments, let me know if they make sense or if we should continue to refine.

**Review by amyevans — APPROVED** [review]
Looks good!

The only remaining thing is the author checklist check is failing, looks like the template was updated recently: https://github.com/Expensify/App/pull/30699/files. @suneox Could you update your checklist to include those new checks please?

**Pull request merged**